### PR TITLE
Feature Pipeline Recipe

### PR DESCRIPTION
**Pull request by tumulurik**

### CODE REVIEW

**Review by kratika2210 — COMMENTED** [review]
@tumulurik We need to update year to 2020 in all the files that use Copyright header 🎉

File: recipes/feature_pipeline_recipes/pyspark/pysparkfeaturepipelineapp/data_saver.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+from pyspark.sql.functions import col, lit, struct
+from .helper import *
+
+
+class MyDatasetSaver(DataSaver):
+
+    def save(self, config_properties, prediction):
+        spark_context = prediction._sc
+        if config_properties is None:
+            raise ValueError("configProperties parameter is null")
```
Comment (ichasepucks) [inline recipes/feature_pipeline_recipes/pyspark/pysparkfeaturepipelineapp/data_saver.py]:
```suggestion
            raise ValueError("config_properties parameter is null")
```

File: recipes/feature_pipeline_recipes/pyspark/pysparkfeaturepipelineapp/data_saver_for_transformed_data.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        df_nested = rename_df.withColumn(tenant_id, f.struct(*([f.col(c).alias(c) for c in rename_df.columns])))\
+                    .drop(*(c for c in rename_df.columns))
+        return df_nested
+
+    def save(self, config_properties, dataframe):
+
+        spark_context = dataframe._sc
+
+        if config_properties is None:
+            raise ValueError("configProperties parameter is null")
```
Comment (ichasepucks) [inline recipes/feature_pipeline_recipes/pyspark/pysparkfeaturepipelineapp/data_saver_for_transformed_data.py]:
```suggestion
            raise ValueError("config_properties parameter is null")
```

File: recipes/feature_pipeline_recipes/pyspark/pysparkfeaturepipelineapp/scoring_data_loader.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+from sdk.data_loader import DataLoader
+from pyspark.sql.functions import col
+
+
+class ScoringDataLoader(DataLoader):
+    logger = setupLogger(__name__)
+
+    def load(self, config_properties, spark):
+        if config_properties is None :
+            raise ValueError("configProperties parameter is null")
```
Comment (ichasepucks) [inline recipes/feature_pipeline_recipes/pyspark/pysparkfeaturepipelineapp/scoring_data_loader.py]:
```suggestion
            raise ValueError("config_properties parameter is null")
```

File: recipes/feature_pipeline_recipes/pyspark/pysparkfeaturepipelineapp/training_data_loader.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+from sdk.data_loader import DataLoader
+from pyspark.sql.functions import col
+
+
+class TrainingDataLoader(DataLoader):
+    logger = setupLogger(__name__)
+
+    def load(self, config_properties, spark):
+        if config_properties is None :
+            raise ValueError("configProperties parameter is null")
```
Comment (ichasepucks) [inline recipes/feature_pipeline_recipes/pyspark/pysparkfeaturepipelineapp/training_data_loader.py]:
```suggestion
            raise ValueError("config_properties parameter is null")
```

File: recipes/feature_pipeline_recipes/pyspark/pysparkfeaturepipelineapp/train_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+import numpy as np
+
+from sdk.pipeline_factory import PipelineFactory
+
+class TrainPipeline(PipelineFactory):
+
+    def apply(self, config_properties):
+        if config_properties is None:
+            raise ValueError("configProperties parameter is null")
```
Comment (ichasepucks) [inline recipes/feature_pipeline_recipes/pyspark/pysparkfeaturepipelineapp/train_pipeline.py]:
```suggestion
            raise ValueError("config_properties parameter is null")
```

**Review by ichasepucks — APPROVED** [review]
Just one trivial logging change that needs to be changed in several places.  Also, this appears to be written for HDI and the 1.x SDK.  That's fine, but we must be aware that this will be out of date and need to be updated for PALM support and 2.x SDK soon.

**Review by bigdataexplorer — APPROVED** [review]
LGTM.

**Pull request merged**